### PR TITLE
feat(test, e2e) test suit for cypress

### PR DESCRIPTION
**Pull request by Sembauke**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Ref before squash:  #41602

*What does this PR do?* 

This reenables [parallelization](https://docs.cypress.io/guides/guides/parallelization#Overview) again, first of all. This pull request makes it possible to dynamically test every challenge. By generating fixtures already provided by the current codebase you can look at that [here](https://github.com/freeCodeCamp/freeCodeCamp/blob/fd53aec348b866614673e9273e9a429c8ed0d212/generate-fixture-data.js). After generating these fixture files (basically just JSON), these files are read by [generate-spec-files.js](https://github.com/freeCodeCamp/freeCodeCamp/blob/fd53aec348b866614673e9273e9a429c8ed0d212/generate-spec-files.js). This ensures the Cypress test are dynamically generated. 
 
The generate spec-files file divides projects and challenges so we can do different test on different challenge types when needed. When everything is done being generated the testing will start. This system will only work on push (all the challenges will be tested by 6 parellized containers).

If you want to run these files locally (which I do not recommend doing). You can start by typing `npm run precypress:gen:fixtures`  after that you should run the following command to generate the tests `npm run precypress:gen:test`. After this you can start testing by typing `npm run cypress:dev:run:full`. This will include other tests aswell outside the testing framwork.

You can add new test to challenges by adding them in the command file provided by Cypress. All tests that are generated will run `cy.testChallenges` or `cy.testPojectsAndBackend` . Back-end challenges which have no editor will be tested by the projects command. 

In the future I recommend to switch to a more sustainable CI such as AWS or Azure to reduce testing time.

### CODE REVIEW

**Review by ojeytonwilliams — COMMENTED** [review]
Hey Sem, awesome job on getting this up and running.  One thing I was wondering: can we avoid specifying the fixtures?  Reason being, all that info is encoded in the curriculum directory, so we shouldn't need to hard code it.  Can't we pull it from the output of `build:curriculum`?

My worry is that this adds a maintenance overhead (every new challenge/project step would require us to update the fixtures).

**Comment by Sembauke** [comment]
> Reason being, all that info is encoded in the curriculum directory, so we shouldn't need to hard code it. Can't we pull it from the output of build:curriculum?

The `cy.updatePaths()` command already pulls path data dynamically from `getChallenges`.  Running a quick `cy.updatePaths('repsonsive-web-design')` should automatically pull and update the path data.

What we can do is adding a command like `precypress:gen:fixtures` dynamically so the fixtures do not get hardcoded into the repo.

**Comment by ojeytonwilliams** [comment]
> What we can do is adding a command like precypress:gen:fixtures dynamically so the fixtures do not get hardcoded into the repo.

Sure, that sounds good.  If that just mirrors the curriculum (respecting `SHOW_UPCOMING_CHANGES`) then that should work perfectly.

**Comment by Sembauke** [comment]
it does not include upcoming changes at the moment. So no upcoming projects are added to the fixture data.

**Comment by ojeytonwilliams** [comment]
> it does not include upcoming changes at the moment. So no upcoming projects are added to the fixture data.

Perfect :+1: 

> before we can do this we should restore the workflows.

In that case, I think it's probably best to try it out on the dev repo.  If everything works (3 containers, all the non-upcoming challenges etc) on the dev repo, then we can implement the workflows as discussed in https://github.com/freeCodeCamp/freeCodeCamp/pull/42011. That way, if something weird happens, no one's PR gets delayed.

File: cypress/plugins/index.js
Comment (ojeytonwilliams) [inline cypress/plugins/index.js:96]: Isn't this all handled by the fixtures now?  Do we still need this?

File: generate-fixture-data.js
Comment (ojeytonwilliams) [inline generate-fixture-data.js:35]:
```suggestion
  mkdirSync(path.join(__dirname, '/cypress/fixtures/path-data'));
  mkdirSync(path.join(__dirname, '/cypress/fixtures/path-data/challenges'));
  mkdirSync(
    path.join(__dirname, '/cypress/fixtures/pathData/projects-and-back-challenges')
```

File: generate-fixture-data.js
Comment (ojeytonwilliams) [inline generate-fixture-data.js:101]:
```suggestion
      `/cypress/fixtures/path-data/challenges/${superblock}.json`
```

File: generate-fixture-data.js
Comment (ojeytonwilliams) [inline generate-fixture-data.js:109]:
```suggestion
      `/cypress/fixtures/path-data/projects-and-back-challenges/${superblock}.json`
```

File: generate-spec-files.js
Comment (ojeytonwilliams) [inline generate-spec-files.js:10]:
```suggestion
    path.join(__dirname, '/cypress/fixtures/path-data/challenges')
```

File: generate-spec-files.js
Comment (ojeytonwilliams) [inline generate-spec-files.js:14]:
```suggestion
    path.join(__dirname, '/cypress/fixtures/path-data/projects-and-back-challenges')
```

File: generate-spec-files.js
Comment (ojeytonwilliams) [inline generate-spec-files.js:35]:
```suggestion
            `/cypress/fixtures/path-data/${
              project ? 'projects-and-back-challenges' : 'challenges'
```

File: generate-spec-files.js
Comment (ojeytonwilliams) [inline generate-spec-files.js:54]:
```suggestion
            const superBlockPath = require('../../../fixtures/path-data/${
              project ? 'projects-and-back-challenges' : 'challenges'
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hi @Sembauke. Thanks for your patience.

One thing jumped out at me: it seems like the plugins are no-longer used.  If that's correct, we can cut out a fair chunk of code.

More detailed comments below:

**Comment by Sembauke** [comment]
Hey @ojeytonwilliams ,
Those are a lot of changes, I will get back to you soon.

File: cypress/support/commands.js
Comment (Sembauke) [reply cypress/support/commands.js:39]: This function was mostly just for precaution. I will delete it.

File: cypress/plugins/index.js
Comment (Sembauke) [reply cypress/plugins/index.js:96]: Yes it is.

File: cypress/integration/challenge-tests/updateTests.js
Comment (ojeytonwilliams) [inline cypress/integration/challenge-tests/updateTests.js:8]: I think this file can go.

File: generate-spec-files.js
Comment (ojeytonwilliams) [inline generate-spec-files.js:31]:
```suggestion
  function divider(files, project) {
```
Nitpick

File: generate-spec-files.js
Comment (ojeytonwilliams) [inline generate-spec-files.js:84]:
```suggestion
  divider(challengesFiles, false);
  divider(projectsFiles, true);
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Sorry about the massive review, but I didn't want to hassle you about minor stuff while you were working.  

Can we gitignore everything in `cypress/fixtures/path-data` and all the js in `cypress/integration/challenge-tests/blocks` ?  Otherwise we might surprise people who generate those files locally.

Aside from that, I made a few more nitpicks, but this is looking solid.

File: cypress/integration/challenge-tests/updateTests.js
Comment (Sembauke) [reply cypress/integration/challenge-tests/updateTests.js:8]: Hmm, thought I deleted that in the last commit, oops.

File: generate-spec-files.js
Comment (Sembauke) [reply generate-spec-files.js:84]: Hahah, English as a second language.

**Comment by Sembauke** [comment]
Hey Oliver,
Before we merge I want to do some more testing locally to see if everything works correctly.

I want to do a full run locally.

**Comment by ojeytonwilliams** [comment]
Sure, @Sembauke, that's totally fine.  Just give me a shout when you're happy for us to proceed.

**Comment by Sembauke** [comment]
It's ready to go.

**Review by ojeytonwilliams — APPROVED** [review]
This is amazing work, @Sembauke, it definitely looks good to me!

I'd be happy with this going in, though obviously someone else should take a look first since I've made commits etc.

**Review by raisedadead — CHANGES_REQUESTED** [review]
Could you remind me why would need two separate workflows?

**Comment by Sembauke** [comment]
> Could you remind me why would need two separate workflows?

It's too expensive to run this system on pull, that's the only reason.

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
Just a quick heads up that these timings are not looking promising. We may end up reducing the frequency to a daily cron job instead.

image

**Comment by Sembauke** [comment]
oh, that does not look good no.

**Comment by raisedadead** [comment]
I went ahead and reverted, more details here: https://github.com/freeCodeCamp/freeCodeCamp/pull/42488#issue-669755364

**Comment by Sembauke** [comment]
So it's probably the workflow because the tests run within 13 minutes.
image

**Comment by raisedadead** [comment]
Yes - Thanks for your work on this. 

> So it's probably the workflow because the tests run within 13 minutes.

We believe so. 

In any case, I think we should dial in the exact behaviour that we want with the other test org. And by that I mean testing the exact code, profiling the run-times. Replicating this repo as close as we can before we are able to integrate it here.

Unfortunately had to revert, for now, to unclog other things that we are working on at the minute. 

Oliver should be able to optimize the workflow and make a subsequent PR when he has gotten a chance - mostly because he is up to speed with the deployment strategy, etc.

**Comment by Sembauke** [comment]
Thanks for the tryout @raisedadead , have been looking forward to this moment 😄 :

**Pull request merged**
